src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+    // set to the beginning of the output term.
+    std::string generate_output_term(const char_int_map& label_count, size_t ellipsis_ndim) const
+    {
+        std::string output_term = ellipsis_ndim == 0 ? "" : "*";
+        for(const auto [label, count] : label_count)
+            if(count == 1)
+                output_term += label;
+
+        return output_term;
+    }
+
+    // Creates a matrix representation of the equation.
+    //
+    // Rows correspond to equation terms, in order of appearance.
+    //
+    // Columns represent the unique labels contained in the equation, ordered alphabetically. If
+    // ellipses are present in the equation, they are represented by the final N columns(N being the
+    // number of dimensions covered by and ellipsis).
+    // Labels not present in a given term are signified by -1.
+    // Labels present in a given term are signified by the input axis they represent.
+    //
+    // e.g. For equation "...ik,kj...->ij...", assuming ... cover two dimensions, the resulting
+    // matrix is:
+    // +-------+----+----+----+---+---+
+    // |       | i  | j  | k  | * | * |
+    // +-------+----+----+----+---+---+
+    // | ...ik |  2 | -1 |  3 | 0 | 1 |
+    // | kj... | -1 |  1 |  0 | 2 | 3 |
+    // | ij... |  0 |  1 | -1 | 2 | 3 |
+    // +-------+----+----+----+---+---+
+    int_mat make_mapping_matrix(const string_vec& terms,
+                                std::string_view unique_labels,
+                                size_t ellipses_ndim) const
+    {
+        char_int_map label_to_column;
+        for(auto i = 0; i < unique_labels.size(); ++i)
+            label_to_column[unique_labels[i]] = i;
+
+        int_mat map_mat = make_matrix(terms.size(), unique_labels.size() + ellipses_ndim, -1);
+
+        for(auto i = 0; i < terms.size(); ++i)
+        {
+            const auto& term = terms[i];
+            int col_id       = 0;
+            for(const auto l : term)
+            {
+                if(l == '*')
+                {
+                    std::iota(map_mat[i].end() - ellipses_ndim, map_mat[i].end(), col_id);
+                    col_id += ellipses_ndim;
+                }
+                else
+                    map_mat[i][label_to_column[l]] = col_id++;
+            }
+        }
+
+        return map_mat;
+    }
+
+    std::vector<char_map<int_vec>> find_duplicates(const string_vec& terms) const
+    {
+        std::vector<char_map<int_vec>> duplicates;
+        for(const auto& term : terms)
+        {
+            char_map<int_vec> counts;
+            for(auto i = 0; i < term.size(); ++i)
+                counts[term[i]].push_back(i);
+
+            for(auto it = counts.begin(); it != counts.end();)
+            {
+                if(it->second.size() < 2)
+                    it = counts.erase(it);
+                else
+                    ++it;
+            }
+
+            duplicates.push_back(counts);
+        }
+
+        return duplicates;
+    }
```

Comment:
It seems like it would be more effecient if we use the `group_uniqe` and `std::sort` algorithm to find the duplicates, especially since you dont even need a map just the counts in sorted order.

```cpp
std::vector<std::vector<int>> find_duplicates(const string_vec& terms) const
{
    std::vector<std::vector<int>> result;
    std::transform(terms.begin(), terms.end(), std::back_inserter(result), [](auto term) {
        std::vector<int> counts;
        std::sort(term.begin(), term.end());
        group_unique(term.begin(), term.end(), [](auto start, auto last) {
            auto n = last-start;
            if (n < 2)
                return;
            counts.push_back(n);
        }, std::equal_to<>{});
        return counts;
    });
    return result;
}
```

Reply:
The naming here is misleading since we don't store the counts, we store the axes on which the duplicated labels occur. I renamed the map to `duplicated_axes` and added a comment explaining the function.

e.g. For equation "iikjj,jkj", the result is a vector containing the two following maps:
result[0]: {'i': [0, 1], 'j': [3, 4]}
result[1]: {'j': [0, 2]}

This is also implemented keeping in mind that in the future we might need to support multiple duplicates per input.